lib/segment/src/index/field_index/geo_index/mutable_geo_index.rs

```diff
@@ -0,0 +1,310 @@
+            }
+
+            if self.point_to_values[idx as usize].is_empty() {
+                self.points_count += 1;
+            }
+
+            points_to_hashes
+                .entry(idx)
+                .or_default()
+                .push(geo_hash.clone());
+
+            self.point_to_values[idx as usize].push(geo_point);
+            self.points_map
+                .entry(geo_hash.clone())
+                .or_default()
+                .insert(idx);
+
+            self.points_values_count += 1;
+        }
+
+        for (_idx, geo_hashes) in points_to_hashes.into_iter() {
+            self.max_values_per_point = max(self.max_values_per_point, geo_hashes.len());
+            self.increment_hash_point_counts(&geo_hashes);
+            for geo_hash in geo_hashes {
+                self.increment_hash_value_counts(&geo_hash);
+            }
+        }
+        Ok(true)
+    }
+
+    pub fn remove_point(&mut self, idx: PointOffsetType) -> OperationResult<()> {
+        if self.point_to_values.len() <= idx as usize {
+            return Ok(()); // Already removed or never actually existed
+        }
+
+        let removed_geo_points = std::mem::take(&mut self.point_to_values[idx as usize]);
+
+        if removed_geo_points.is_empty() {
+            return Ok(());
+        }
+
+        self.points_count -= 1;
+        self.points_values_count -= removed_geo_points.len();
+        let mut removed_geo_hashes = Vec::with_capacity(removed_geo_points.len());
+
+        for removed_geo_point in removed_geo_points {
+            let removed_geo_hash: GeoHash =
+                encode_max_precision(removed_geo_point.lon, removed_geo_point.lat).unwrap();
+            removed_geo_hashes.push(removed_geo_hash.clone());
+
+            let key = GeoMapIndex::encode_db_key(&removed_geo_hash, idx);
+            self.db_wrapper.remove(key)?;
+
+            let is_last = if let Some(hash_ids) = self.points_map.get_mut(&removed_geo_hash) {
+                hash_ids.remove(&idx);
+                hash_ids.is_empty()
+            } else {
+                log::warn!(
+                    "Geo index error: no points for hash {} was found",
+                    removed_geo_hash
+                );
+                false
+            };
+
+            if is_last {
+                self.points_map.remove(&removed_geo_hash);
+            }
+
+            self.decrement_hash_value_counts(&removed_geo_hash);
+        }
+
+        self.decrement_hash_point_counts(&removed_geo_hashes);
+        Ok(())
+    }
+
+    pub fn add_many_geo_points(
+        &mut self,
+        idx: PointOffsetType,
+        values: &[GeoPoint],
+    ) -> OperationResult<()> {
+        if values.is_empty() {
+            return Ok(());
+        }
+
+        if self.point_to_values.len() <= idx as usize {
+            // That's a smart reallocation
+            self.point_to_values.resize_with(idx as usize + 1, Vec::new);
+        }
```

Comment:
If we, as user, insert points one by one. Do we call this branch every time? In that case, this would actually be quite inefficient. It would then be better to grow by a factor of 2 to prevent constant expensive reallocations.

Reply:
You're totally right! Thank you for confirming.